### ISSUE TITLE
What's preventing vkEvent to cross queues?

### ISSUE THREAD

**Issue by ShabbyX** [KhronosGroup/Vulkan-Docs#771]
On PS4, you can use a Label (very similar to vkEvent) to signal an event and wait on it. This can cross any queues (which by Vulkan specifications would belong to two different families, as one is graphics-compute and the rest are compute-only). This is GCN, so at least some current hardware is capable of doing this.

This makes multi-queue rendering very efficient as events between the graphics queue and the async compute queues can be very frequent, allowing detailed layout of parallel work where they make sense (for example overlap of memory-bound and VALU bound, or vertex bound and compute). With Vulkan, cross-queue synchronization would require a semaphore, which entails expensive submissions.

What's the technical reason why Vulkan doesn't allow a vkEvent to be used to synchronize two queues of the same family? If there are none, what about any two queues?

**Comment by TomOlson** [comment]
> What's the technical reason why Vulkan doesn't allow a vkEvent to be used to synchronize two queues of the same family?

Bottom line, it's hardware / platform variability. Vulkan queues are very much an abstraction - they map nicely onto everyone's hardware, but the nature of the mapping varies a lot across vendors. Lightweight cross-queue synchronization in mid-command-buffer would be hugely invasive to some drivers, and the kinds of scheduling optimizations it would enable probably wouldn't be particularly performance-portable. A separate issue, I'm told (don't know this for a fact myself) that there are platform constraints e.g. on Windows, that limit the app/driver's ability to do dynamic scheduling across multiple queues. Thus it would be easy to set up deadlock situations.

It does work on GCN, so I suppose AMD might be able to expose an extension, if the platform issues aren't a blocker. Would that be useful?

**Comment by ShabbyX** [comment]
Thanks for the explanation Tom. If Linux doesn't have such a constraint, then yes, an extension would definitely be useful, at least for a few upcoming announced and unannounced Linux/AMD based console-likes.

**Comment by TomOlson** [comment]
@ShabbyX, we discussed again today - feedback was that there are problems on both Windows and Linux, though not the _same_ problems. Result is that we aren't likely to tackle it as a cross-vendor extension any time soon. It sounds like you have the right contacts to raise this with AMD directly, and that's probably your best course. Consoles even if virtual have more room to maneuver than we do, so there may be workarounds they can take that wouldn't work for the general case.

Since we're not going to work on this, I'm going to close the issue, but thanks for raising it. We'll be watching to see what happens in the console-like space.
